pkg/local-pv-monitor/monitor.go

```diff
@@ -0,0 +1,508 @@
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", obj)
+		return
+	}
+
+	monitor.localVolumeMap.DeleteLocalVolume(volume)
+
+}
+
+// Run starts all of this controller's control loops
+func (monitor *LocalPVMonitor) Run(stopCh <-chan struct{}) {
+	glog.Infof("Starting monitor controller %s!", string(monitor.RuntimeConfig.Name))
+	monitor.hasRunLock.Lock()
+	monitor.hasRun = true
+	monitor.hasRunLock.Unlock()
+	go monitor.volumeController.Run(stopCh)
+
+	go monitor.MonitorLocalVolumes()
+	<-stopCh
+}
+
+// HasRun returns whether the volume controller has Run
+func (monitor *LocalPVMonitor) HasRun() bool {
+	monitor.hasRunLock.Lock()
+	defer monitor.hasRunLock.Unlock()
+	return monitor.hasRun
+}
+
+// MonitorLocalVolumes checks local PVs periodically
+func (monitor *LocalPVMonitor) MonitorLocalVolumes() {
+	for {
+		if monitor.HasRun() {
+			pvs := monitor.localVolumeMap.GetPVs()
+			for _, pv := range pvs {
+				monitor.checkStatus(pv)
+			}
+		}
+
+		time.Sleep(DefaultMonitorResyncPeriod)
+	}
+}
+
+// checkStatus checks pv health condition
+func (monitor *LocalPVMonitor) checkStatus(pv *v1.PersistentVolume) {
+	// check if PV is local storage
+	if pv.Spec.Local == nil {
+		glog.Infof("PV: %s is not local storage", pv.Name)
+		return
+	}
+	// check node and pv affinity
+	fit, err := CheckNodeAffinity(pv, monitor.Node.Labels)
+	if err != nil {
+		glog.Errorf("check node affinity error: %v", err)
+		return
+	}
+	if !fit {
+		glog.Errorf("pv: %s does not belong to this node: %s", pv.Name, monitor.Node.Name)
+		return
+	}
+
+	// check if host dir still exists
+	mountPath, continueThisCheck := monitor.checkHostDir(pv)
+	if !continueThisCheck {
+		glog.Errorf("Host dir is modified, PV should be marked")
+		return
+	}
+
+	// check if it is still a mount point
+	continueThisCheck = monitor.checkMountPoint(mountPath, pv)
+	if !continueThisCheck {
+		glog.Errorf("Retrieving mount points error or %s is not a mount point any more", mountPath)
+		return
+	}
+
+	// check PV size: PV capacity must not be greater than device capacity and PV used bytes must not be greater that PV capacity
+	dir, _ := monitor.VolUtil.IsDir(mountPath)
+	if dir {
+		monitor.checkPVAndFSSize(mountPath, pv)
+	}
+	bl, _ := monitor.VolUtil.IsBlock(mountPath)
+	if bl {
+		monitor.checkPVAndBlockSize(mountPath, pv)
+	}
+}
+
+func (monitor *LocalPVMonitor) checkMountPoint(mountPath string, pv *v1.PersistentVolume) bool {
+	// Retrieve list of mount points to iterate through discovered paths (aka files) below
+	mountPoints, mountPointsErr := monitor.RuntimeConfig.Mounter.List()
+	if mountPointsErr != nil {
+		glog.Errorf("Error retrieving mount points: %v", mountPointsErr)
+		return false
+	}
+	// Check if mountPath is still a mount point
+	for _, mp := range mountPoints {
+		if mp.Path == mountPath {
+			glog.V(10).Infof("mountPath is still a mount point: %s", mountPath)
+			err := monitor.markOrUnmarkPV(pv, NotMountPoint, "yes", false)
```

Comment:
if i change from a mountpoint to not a mountpoint, then change it back again, will there be any problem?

Reply:
yeah, i am thinking this too.
Basically i think we still need to mark the local PV as unhealthy, because it had been not a mountpoint for a period of time (users may have lost some data).
But i will unmark the local pv in the implementation now. Maybe i need to change it